src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt

```diff
@@ -133,13 +133,64 @@ class RsErrorAnnotator : AnnotatorBase(), HighlightRangeExtension {
 
     private fun checkUseSpeck(holder: RsAnnotationHolder, useSpeck: RsUseSpeck) {
         if (useSpeck.isStarImport || useSpeck.useGroup != null) return
+
+        checkDuplicateImport(holder, useSpeck)
+        checkReexports(holder, useSpeck)
+    }
+
+    private fun checkDuplicateImport(holder: RsAnnotationHolder, useSpeck: RsUseSpeck) {
         val duplicates = holder.currentAnnotationSession.duplicatesByNamespace(useSpeck.containingMod, false)
         if (useSpeck.namespaces.any { useSpeck in duplicates[it].orEmpty() }) {
             val identifier = PsiTreeUtil.getDeepestLast(useSpeck)
             RsDiagnostic.DuplicateImportError(identifier).addToHolder(holder)
         }
     }
 
+    private fun checkReexports(holder: RsAnnotationHolder, useSpeck: RsUseSpeck) {
+        val item = useSpeck.ancestorStrict<RsUseItem>() ?: return
+        if (item.visibility == RsVisibility.Private) return
+
+        val path = useSpeck.path ?: return
+        val targets = path.reference?.multiResolve()?.filterIsInstance<RsItemElement>() ?: emptyList()
+
+        // targets not accessible from the use element will be handled by E0603
+        val visibleTargets = targets.filter { it.isVisibleFrom(useSpeck.containingMod) }
+        val invalidTargets = visibleTargets.filter {
+            !canBeReexported(item.visibility, it)
+        }
+
+        // if at least one item can be re-exported or nothing is invalid, allow the reexport
+        if (invalidTargets.size < targets.size || invalidTargets.isEmpty()) return
+
+        val invalidTarget = invalidTargets.first()
+        val context = useSpeck.context
+        val name = if (path.referenceName == "self" && context is RsUseGroup) {
+            (context.context as? RsUseSpeck)?.path?.referenceName ?: path.referenceName
+        } else {
+            path.referenceName
+        }
+
+        RsDiagnostic.InvalidReexport(path.referenceNameElement, name, invalidTarget).addToHolder(holder)
+    }
+
+    private fun canBeReexported(reexportVisibility: RsVisibility, item: RsItemElement): Boolean {
+        return when (val itemVisibility = item.visibility) {
+            is RsVisibility.Private ->
+                when (reexportVisibility) {
+                    is RsVisibility.Private -> true
+                    is RsVisibility.Restricted -> reexportVisibility.inMod == item.containingMod
```

Comment:
I support it still a cause of false-positives
```rust
mod foo {
    mod bar {}
    mod baz {
        pub(in crate::foo::baz) use crate::foo::bar as qwe;
    }
}
```


![image](https://user-images.githubusercontent.com/2539310/97455422-e2e61000-1948-11eb-8988-cfdc000ded65.png)

Reply:
Take #2 :D Now it should be hopefully general. I understood private visibility wrong before.